tests/test_spi_devices.py

```diff
@@ -0,0 +1,339 @@
+
+    def on_start(self):
+        super(MockMCP3xxx, self).on_start()
+        self.state = 'idle'
+
+    def on_bit(self):
+        if self.state == 'idle':
+            if self.rx_buf[-1]:
+                self.state = 'mode'
+                self.rx_buf = []
+        elif self.state == 'mode':
+            if self.rx_buf[-1]:
+                self.state = 'single'
+            else:
+                self.state = 'diff'
+            self.rx_buf = []
+        elif self.state in ('single', 'diff'):
+            if len(self.rx_buf) == self.channel_bits:
+                self.on_result(self.state == 'diff', self.rx_word())
+                self.state = 'result'
+        elif self.state == 'result':
+            if not self.tx_buf:
+                self.state = 'idle'
+                self.rx_buf = []
+        else:
+            assert False
+
+    def on_result(self, differential, channel):
+        if differential:
+            pos_channel = channel
+            neg_channel = pos_channel ^ 1
+            result = self.channels[pos_channel] - self.channels[neg_channel]
+            result = clamp(result, 0, self.vref)
+        else:
+            result = clamp(self.channels[channel], 0, self.vref)
+        result = scale(result, self.vref, self.bits)
+        self.tx_word(result, self.bits + 2)
+
+
+class MockMCP3xx1(MockMCP3xxx):
+    def __init__(self, clock_pin, mosi_pin, miso_pin, select_pin=None, bits=10):
+        super(MockMCP3xx1, self).__init__(
+            clock_pin, mosi_pin, miso_pin, select_pin, channels=2, bits=bits)
+
+    def on_start(self):
+        super(MockMCP3xx1, self).on_start()
+        result = self.channels[0] - self.channels[1]
+        result = clamp(result, 0, self.vref)
+        result = scale(result, self.vref, self.bits)
+        self.tx_word(result, self.bits + 3)
+
+    def on_bit(self):
+        pass
+
+
+class MockMCP3xx2(MockMCP3xxx):
+    def __init__(
+            self, clock_pin, mosi_pin, miso_pin, select_pin=None,
+            bits=10):
+        super(MockMCP3xx2, self).__init__(
+            clock_pin, mosi_pin, miso_pin, select_pin, channels=2, bits=bits)
+        self.channel_bits = 1
+
+
+class MockMCP33xx(MockMCP3xxx):
+    def __init__(
+            self, clock_pin, mosi_pin, miso_pin, select_pin=None,
+            channels=8):
+        super(MockMCP33xx, self).__init__(
+            clock_pin, mosi_pin, miso_pin, select_pin, channels, 12)
+
+    def on_result(self, differential, channel):
+        if differential:
+            pos_channel = channel
+            neg_channel = pos_channel ^ 1
+            result = self.channels[pos_channel] - self.channels[neg_channel]
+            result = clamp(result, -self.vref, self.vref)
+        else:
+            result = clamp(self.channels[channel], 0, self.vref)
+        result = scale(result, self.vref, self.bits)
+        if result < 0:
+            result += 8192
+        self.tx_word(result, self.bits + 3)
+
+
+class MockMCP3001(MockMCP3xx1):
```

Comment:
All these MockMCP classes should probably be refactored to match what I did in #281